File: app/src/main/java/com/eagskunst/emmanuel/gamingnews/utility/di/IgdbModule.java
package com.eagskunst.emmanuel.gamingnews.utility.di;

import com.eagskunst.emmanuel.gamingnews.api.GamesApi;
import com.eagskunst.emmanuel.gamingnews.models.Cover;
import com.eagskunst.emmanuel.gamingnews.utility.CoverConverter;
import com.google.gson.Gson;

import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by eagskunst on 11/01/2019
 */
@Module
public class IgdbModule {

    @Provides
    @IgbScope
    public Retrofit provideRetrofit(){
        Gson gson = new Gson().newBuilder()
                .registerTypeAdapter(Cover.class, new CoverConverter())
                .create();
        GsonConverterFactory gsonConverterFactory = GsonConverterFactory.create(gson);
        return new Retrofit.Builder()
                .baseUrl("https://api-v3.igdb.com/")
                .addConverterFactory(gsonConverterFactory)
                .build();
    }

    @Provides
    @IgbScope
    public GamesApi provideApi(Retrofit retrofit){
        return retrofit.create(GamesApi.class);
    }
}

File: app/src/main/java/com/eagskunst/emmanuel/gamingnews/utility/di/IgdbModule.kt
package com.eagskunst.emmanuel.gamingnews.utility.di

import com.eagskunst.emmanuel.gamingnews.api.GamesApi
import com.eagskunst.emmanuel.gamingnews.models.Cover
import com.eagskunst.emmanuel.gamingnews.utility.CoverConverter
import com.google.gson.Gson
import dagger.Module
import dagger.Provides
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

/**
 * Created by eagskunst on 11/01/2019
 */
@Module
class IgdbModule {

    @Provides
    @IgbScope
    fun provideRetrofit(): Retrofit {
        val gson = Gson().newBuilder()
            .registerTypeAdapter(Cover::class.java, CoverConverter())
            .create()
        val gsonConverterFactory = GsonConverterFactory.create(gson)
        return Retrofit.Builder()
            .baseUrl("https://api-v3.igdb.com/")
            .addConverterFactory(gsonConverterFactory)
            .build()
    }

    @Provides
    @IgbScope
    fun provideApi(retrofit: Retrofit): GamesApi {
        return retrofit.create(GamesApi::class.java)
    }
}
